src/lib/ota/common.ts

```diff
@@ -427,265 +508,242 @@ export async function isNewImageAvailable(current: Ota.ImageInfo, device: Zh.Dev
-                                'genOta', 'imageBlockResponse', payload, null, transactionSequenceNumber,
-                            ).then(
-                                () => {
-                                    pageOffset += payload.dataSize;
-                                    lastImageBlockResponse = Date.now();
-                                    thenCallback();
-                                },
-                                (e) => {
-                                    // Shit happens, device will probably do a new imageBlockRequest so don't care.
-                                    lastImageBlockResponse = Date.now();
-                                    thenCallback();
-                                    logger.debug(`Image block response failed (${e.message})`, NS);
-                                },
-                            );
-                        }, cooldownTime);
-
-                        lastUpdate = callOnProgress(startTime, lastUpdate, imageBlockRequest, image, onProgress);
-                    };
-
-                    if ('pageSize' in imageBlockOrPageRequest.payload) {
-                        // imagePageRequest
-                        pageSize = imageBlockOrPageRequest.payload.pageSize as number;
-                        const handleImagePageRequestBlocks = (imagePageRequest: KeyValueAny) => {
-                            if (pageOffset < pageSize) {
-                                sendImageBlockResponse(imagePageRequest,
-                                    () => handleImagePageRequestBlocks(imagePageRequest), imagePageRequest.header.transactionSequenceNumber);
-                            } else {
-                                answerNextImageBlockOrPageRequest();
-                            }
-                        };
-                        handleImagePageRequestBlocks(imageBlockOrPageRequest);
-                    } else {
-                        // imageBlockRequest
-                        sendImageBlockResponse(imageBlockOrPageRequest, answerNextImageBlockOrPageRequest,
-                            imageBlockOrPageRequest.header.transactionSequenceNumber);
-                    }
-                },
-                () => {
-                    cancelWaiters(waiters);
-                    reject(new Error(`OTA: Timeout, device did not request any image blocks`));
-                },
-            );
+        const imageBlockRequest = endpoint.waitForCommand('genOta', 'imageBlockRequest', null, imageBlockOrPageRequestTimeoutMs);
+        const imagePageRequest = endpoint.waitForCommand('genOta', 'imagePageRequest', null, imageBlockOrPageRequestTimeoutMs);
+        waiters.imageBlockOrPageRequest = {
+            promise: Promise.race([imageBlockRequest.promise, imagePageRequest.promise]),
+            cancel: () => {
+                imageBlockRequest.cancel();
+                imagePageRequest.cancel();
+            },
         };
 
-        const answerNextImageRequest = () => {
```

Comment:
I see this code was removed without a replacement. I'm not sure why it was there though because the `requestOTA` is already done before it, so this would have never got the `queryNextImageRequest`

Reply:
`requestOTA` doesn't respond to `queryNextImageRequest`, which the device needs to continue and start asking for image blocks. That was duplicated with that function, but with the response. Now it's just called directly with `sendQueryNextImageResponse` after `getNewImage`.
_That was probably also responsible for the (sometimes) very bad ETA on start._